File: php/tests/conformance.php
<?php

/**
 * The 34-check binding conformance suite (docs/conformance.md, v2).
 *
 * Proves the PHP binding marshals every value shape across the FFI. It is NOT a
 * phone-number test suite — the behavioural cases live in the engine's own
 * tests and run once, in Aether.
 *
 * ## Why a plain runner and not PHPUnit
 *
 * PHPUnit arrives via Composer, so `vendor/bin/phpunit` needs a `composer
 * install` — a network round trip, or a pre-warmed cache — before a single
 * assertion executes. The rest of this monorepo's bindings test with whatever
 * is already on the box, so this one does too: `php tests/conformance.php`, no
 * dependencies, and the process exit code is the result.
 *
 *     LIBPHONENUMBER_AE_LIB=../target/build/core/lib/libphonenumber_ae.so \
 *         php -d ffi.enable=1 tests/conformance.php
 */

declare(strict_types=1);

// Composer's autoloader when the package has been installed; otherwise a
// four-line PSR-4 stand-in, so the suite runs from a bare checkout with no
// `composer install` (and therefore no network).
if (is_file(__DIR__ . '/../vendor/autoload.php')) {
    require __DIR__ . '/../vendor/autoload.php';
} else {
    spl_autoload_register(static function (string $class): void {
        $prefix = 'PhoneNumberAe\\';
        if (!str_starts_with($class, $prefix)) {
            return;
        }
        $file = __DIR__ . '/../src/' . substr($class, strlen($prefix)) . '.php';
        if (is_file($file)) {
            require $file;
        }
    });
}

use PhoneNumberAe\AsYouTypeFormatter;
use PhoneNumberAe\PhoneNumber;

$passed = 0;
/** @var list<string> $failures */
$failures = [];

/** Run one check; $body throws on failure. */
function check(string $name, callable $body): void
{
    global $passed, $failures;
    try {
        $body();
        $passed++;
        echo "  PASS {$name}\n";
    } catch (\Throwable $e) {
        $failures[] = "{$name}: {$e->getMessage()}";
        echo "  FAIL {$name}\n";
        echo "       {$e->getMessage()}\n";
    }
}

function eqStr(string $got, string $want, string $what = 'value'): void
{
    if ($got !== $want) {
        throw new \Exception(sprintf("%s:\n         got  \"%s\"\n         want \"%s\"", $what, $got, $want));
    }
}

function eqInt(int $got, int $want, string $what = 'value'): void
{
    if ($got !== $want) {
        throw new \Exception("{$what}: got {$got}, want {$want}");
    }
}

function isTrue(bool $got, string $what): void
{
    if (!$got) {
        throw new \Exception("{$what}: expected true");
    }
}

function isFalse(bool $got, string $what): void
{
    if ($got) {
        throw new \Exception("{$what}: expected false");
    }
}

function atLeast(int $got, int $min, string $what): void
{
    if ($got < $min) {
        throw new \Exception("{$what}: got {$got}, want >= {$min}");
    }
}

echo "=== phonenumber_ae PHP binding conformance (v2) ===\n";
if (!extension_loaded('ffi')) {
    fwrite(STDERR, "php: ext-ffi is not loaded\n");
    exit(2);
}
printf(
    "engine: %s (ABI v%d)\n",
    PhoneNumber::nativeLibraryPath() ?? '(unknown)',
    PhoneNumber::abiVersion()
);

// ---- the thirty-four ----

check('01 country_code US', function (): void {
    eqStr(PhoneNumber::countryCode('US'), '1');
});

check('02 country_code GB', function (): void {
    eqStr(PhoneNumber::countryCode('GB'), '44');
});

check('03 unknown region', function (): void {
    eqStr(PhoneNumber::countryCode('ZZ'), '');
});

check('04 example_number US', function (): void {
    eqStr(PhoneNumber::exampleNumber('US'), '2015550123');
});

check('05 possible_lengths US', function (): void {
    eqStr(PhoneNumber::possibleLengths('US'), '10');
});

check('06 region_code_for_country_code 44', function (): void {
    eqStr(PhoneNumber::regionCodeForCountryCode('44'), 'GB');
});

check('07 is_nanpa_country US', function (): void {
    isTrue(PhoneNumber::isNanpaCountry('US'), 'is_nanpa US');
});

check('08 region enumeration', function (): void {
    $regs = PhoneNumber::regions();
    atLeast(count($regs), 200, 'region count');
    eqInt(strlen($regs[0]), 2, 'region_at(0) length');
});

check('09 cc_region_at 1,0', function (): void {
    eqStr(PhoneNumber::regionsForCountryCode('1')[0], 'US');
});

check('10-14 parse', function (): void {
    $num = PhoneNumber::parse('+1 201 555 0123 ext 42', 'US');
    eqStr($num->error(), '', 'parse error');
    eqStr($num->nationalNumber(), '2015550123', 'national_number');       // 10
    eqStr($num->extension(), '42', 'extension');                          // 11
    eqStr($num->countryCode(), '1', 'country_code');                     // 12
    eqInt($num->source(), PhoneNumber::SRC_FROM_NUMBER_WITH_PLUS, 'source'); // 13
    eqStr($num->regionCode(), 'US', 'region_code');                      // 14
});

check('15 parse trunk-prefix strip (GB)', function (): void {
    eqStr(PhoneNumber::parse('01212345678', 'GB')->nationalNumber(), '1212345678');
});

check('16 is_possible yes', function (): void {
    isTrue(PhoneNumber::isPossibleNumber('US', '2015550123'), 'is_possible yes');
});

check('17 reason too short', function (): void {
    eqInt(PhoneNumber::isPossibleNumberWithReason('US', '201555'), PhoneNumber::VR_TOO_SHORT, 'reason');
});

check('18 is_valid yes', function (): void {
    isTrue(PhoneNumber::isValidNumber('US', '2015550123'), 'is_valid yes');
});

check('19 is_valid wrong shape', function (): void {
    isFalse(PhoneNumber::isValidNumber('US', '1015550123'), 'is_valid wrong shape');
});

check('20 is_valid with +cc', function (): void {
    isTrue(PhoneNumber::isValidNumber('US', '+12015550123'), 'is_valid with +cc');
});

check('21 number_type fixed_line_or_mobile', function (): void {
    // US fixedLine==mobile -> FIXED_LINE_OR_MOBILE; GB has distinct patterns
    eqInt(PhoneNumber::numberType('US', '2015550123'), PhoneNumber::TYPE_FIXED_LINE_OR_MOBILE, 'number_type US');
    eqInt(PhoneNumber::TYPE_FIXED_LINE_OR_MOBILE, 10, 'TYPE_FIXED_LINE_OR_MOBILE constant');
    eqInt(PhoneNumber::numberType('GB', '2070313000'), PhoneNumber::TYPE_FIXED_LINE, 'number_type GB');
    eqInt(PhoneNumber::TYPE_FIXED_LINE, 0, 'TYPE_FIXED_LINE constant');
});

check('22 format national', function (): void {
    eqStr(PhoneNumber::format('US', '2015550123', PhoneNumber::NATIONAL), '(201) 555-0123');
});

check('23 format E164', function (): void {
    eqStr(PhoneNumber::format('US', '2015550123', PhoneNumber::E164), '+12015550123');
    eqInt(PhoneNumber::E164, 0, 'E164 constant');
});

check('24 format international', function (): void {
    eqStr(PhoneNumber::format('US', '2015550123', PhoneNumber::INTERNATIONAL), '+1 201-555-0123');
});

check('25 format RFC3966', function (): void {
    eqStr(PhoneNumber::format('US', '2015550123', PhoneNumber::RFC3966), 'tel:+1-201-555-0123');
});

check('26 is_number_match exact', function (): void {
    eqInt(PhoneNumber::isNumberMatch('+12015550123', '+1 201 555 0123'), PhoneNumber::MATCH_EXACT, 'match exact');
});

check('27 is_number_match none', function (): void {
    eqInt(PhoneNumber::isNumberMatch('+12015550123', '+12025550123'), PhoneNumber::MATCH_NO_MATCH, 'match none');
});

check('28 normalize_digits_only', function (): void {
    eqStr(PhoneNumber::normalizeDigitsOnly('+1 (201) 555.0123'), '12015550123');
});

check('29 convert_alpha_characters', function (): void {
    eqStr(PhoneNumber::convertAlphaCharacters('1-800-FLOWERS'), '1-800-3569377');
});

check('30 truncate_too_long', function (): void {
    eqStr(PhoneNumber::truncateTooLong('US', '20155501239999'), '2015550123');
});

check('31 as-you-type', function (): void {
    $ayt = new AsYouTypeFormatter('US');
    $out = '';
    foreach (str_split('2015550123') as $c) {
        $out = $ayt->inputDigit($c);
    }
    eqStr($out, '(201) 555-0123');
});

check('32 matcher count', function (): void {
    $matches = PhoneNumber::findNumbers('call 201-555-0123 or +1 202 555 0199', 'US', PhoneNumber::LENIENCY_VALID);
    eqInt(count($matches), 2, 'matcher count');
});

check('33 matcher raw', function (): void {
    $matches = PhoneNumber::findNumbers('call 201-555-0123 now', 'US', PhoneNumber::LENIENCY_VALID);
    eqStr($matches[0]->raw, '201-555-0123');
});

check('34 abi version', function (): void {
    eqInt(PhoneNumber::abiVersion(), 2, 'abi version');
});

// ---- a few extras exercising the idiomatic surface ----

check('format helpers agree with format()', function (): void {
    eqStr(PhoneNumber::formatNational('US', '2015550123'), '(201) 555-0123');
    eqStr(PhoneNumber::formatInternational('US', '2015550123'), '+1 201-555-0123');
    eqStr(PhoneNumber::formatE164('US', '2015550123'), '+12015550123');
    eqStr(PhoneNumber::formatRfc3966('US', '2015550123'), 'tel:+1-201-555-0123');
});

check('type constants are the documented ints', function (): void {
    eqInt(PhoneNumber::TYPE_UNKNOWN, -1, 'TYPE_UNKNOWN');
    eqInt(PhoneNumber::TYPE_MOBILE, 1, 'TYPE_MOBILE');
    eqInt(PhoneNumber::TYPE_VOICEMAIL, 9, 'TYPE_VOICEMAIL');
});

check('matcher reports start/end offsets', function (): void {
    $matches = PhoneNumber::findNumbers('call 201-555-0123 now', 'US', PhoneNumber::LENIENCY_VALID);
    eqInt($matches[0]->start, 5, 'match start');
    eqInt($matches[0]->end, 17, 'match end');
});

check('many calls do not leak or crash', function (): void {
    // A returned char* that was never freed would show up here as steadily
    // growing RSS; a double free would crash. Cheap insurance over the
    // takeString contract.
    for ($i = 0; $i < 5000; $i++) {
        PhoneNumber::format('US', '2015550123', PhoneNumber::NATIONAL);
    }
    eqStr(PhoneNumber::format('US', '2015550123', PhoneNumber::NATIONAL), '(201) 555-0123');
});

// ---- result ----

printf("=== %d passed, %d failed ===\n", $passed, count($failures));
if ($failures !== []) {
    echo "failures:\n";
    foreach ($failures as $f) {
        echo "  - {$f}\n";
    }
    exit(1);
}
exit(0);

File: rust/tests/conformance.rs
//! The binding conformance suite (docs/conformance.md, v6 — 45 checks).
//!
//! Proves the Rust binding marshals every value shape across the FFI. It is
//! NOT a phone-number test suite — the behavioural cases live in the engine's
//! own tests and run once, in Aether.
//!
//! Each test loads its own engine over `$LIBPHONENUMBER_AE_LIB`, mirroring the
//! other bindings' suites. A couple of extras exercise the typed idiomatic
//! surface (enums, the `PhoneNumbers` object, the `AsYouTypeFormatter` class,
//! the matcher iterator) that still bottoms out at these calls.

use phonenumber_ae as pn;
use phonenumber_ae::PhoneNumbers;

fn engine() -> PhoneNumbers {
    PhoneNumbers::new().expect("load the engine (set LIBPHONENUMBER_AE_LIB)")
}

#[test]
fn t01_country_code_us() {
    assert_eq!(engine().country_code("US"), "1");
}

#[test]
fn t02_country_code_gb() {
    assert_eq!(engine().country_code("GB"), "44");
}

#[test]
fn t03_unknown_region() {
    assert_eq!(engine().country_code("ZZ"), "");
}

#[test]
fn t04_example_number() {
    assert_eq!(engine().example_number("US"), "2015550123");
}

#[test]
fn t05_possible_lengths() {
    assert_eq!(engine().possible_lengths("US"), "10");
}

#[test]
fn t06_region_for_cc() {
    assert_eq!(engine().region_code_for_country_code("44"), "GB");
}

#[test]
fn t07_is_nanpa() {
    assert!(engine().is_nanpa_country("US"));
}

#[test]
fn t08_region_enumeration() {
    let pn = engine();
    assert!(pn.region_count() >= 200, "region_count={}", pn.region_count());
    assert_eq!(pn.region_at(0).len(), 2, "region_at(0)={:?}", pn.region_at(0));
}

#[test]
fn t09_cc_region() {
    assert_eq!(engine().regions_for_country_code("1")[0], "US");
}

#[test]
fn t10_to_14_parse() {
    let pn = engine();
    let num = pn.parse("+1 201 555 0123 ext 42", "US");
    assert_eq!(num.error(), "");
    assert_eq!(num.national_number(), "2015550123"); // #10
    assert_eq!(num.extension(), "42"); // #11
    assert_eq!(num.country_code(), "1"); // #12
    assert_eq!(num.source(), pn::SRC_FROM_NUMBER_WITH_PLUS); // #13
    assert_eq!(num.region_code(), "US"); // #14
}

#[test]
fn t15_parse_trunk_prefix() {
    assert_eq!(
        engine().parse("01212345678", "GB").national_number(),
        "1212345678"
    );
}

#[test]
fn t16_is_possible() {
    assert!(engine().is_possible_number("US", "2015550123"));
}

#[test]
fn t17_reason_too_short() {
    assert_eq!(
        engine().is_possible_number_with_reason("US", "201555"),
        pn::VR_TOO_SHORT
    );
}

#[test]
fn t18_is_valid() {
    assert!(engine().is_valid_number("US", "2015550123"));
}

#[test]
fn t19_invalid_shape() {
    assert!(!engine().is_valid_number("US", "1015550123"));
}

#[test]
fn t20_valid_with_cc() {
    assert!(engine().is_valid_number("US", "+12015550123"));
}

#[test]
fn t21_number_type() {
    // US fixedLine==mobile -> FIXED_LINE_OR_MOBILE; GB has distinct patterns.
    assert_eq!(
        engine().number_type("US", "2015550123"),
        pn::TYPE_FIXED_LINE_OR_MOBILE
    );
    assert_eq!(engine().number_type("GB", "2070313000"), pn::TYPE_FIXED_LINE);
}

#[test]
fn t22_format_national() {
    assert_eq!(
        engine().format("US", "2015550123", pn::NATIONAL),
        "(201) 555-0123"
    );
}

#[test]
fn t23_format_e164() {
    assert_eq!(
        engine().format("US", "2015550123", pn::E164),
        "+12015550123"
    );
}

#[test]
fn t24_format_international() {
    assert_eq!(
        engine().format("US", "2015550123", pn::INTERNATIONAL),
        "+1 201-555-0123"
    );
}

#[test]
fn t25_format_rfc3966() {
    assert_eq!(
        engine().format("US", "2015550123", pn::RFC3966),
        "tel:+1-201-555-0123"
    );
}

#[test]
fn t26_match_exact() {
    assert_eq!(
        engine().is_number_match("+12015550123", "+1 201 555 0123"),
        pn::MATCH_EXACT
    );
}

#[test]
fn t27_match_none() {
    assert_eq!(
        engine().is_number_match("+12015550123", "+12025550123"),
        pn::MATCH_NO_MATCH
    );
}

#[test]
fn t28_normalize() {
    assert_eq!(
        engine().normalize_digits_only("+1 (201) 555.0123"),
        "12015550123"
    );
}

#[test]
fn t29_alpha() {
    assert_eq!(
        engine().convert_alpha_characters("1-800-FLOWERS"),
        "1-800-3569377"
    );
}

#[test]
fn t30_truncate() {
    assert_eq!(
        engine().truncate_too_long("US", "20155501239999"),
        "2015550123"
    );
}

#[test]
fn t31_as_you_type() {
    let pn = engine();
    let mut ayt = pn.as_you_type_formatter("US");
    let mut out = String::new();
    for c in "2015550123".chars() {
        out = ayt.input_digit(c);
    }
    assert_eq!(out, "(201) 555-0123");
}

#[test]
fn t32_matcher_count() {
    let matches = engine().find_numbers(
        "call 201-555-0123 or +1 202 555 0199",
        "US",
        pn::LENIENCY_VALID,
    );
    assert_eq!(matches.len(), 2);
}

#[test]
fn t33_matcher_raw() {
    let matches = engine().find_numbers("call 201-555-0123 now", "US", pn::LENIENCY_VALID);
    assert_eq!(matches[0].raw, "201-555-0123");
}

#[test]
fn t34_abi_version() {
    assert_eq!(engine().abi_version(), 6);
}

#[test]
fn t35_short_emergency_us() {
    assert!(engine().is_emergency_number("US", "911"));
}

#[test]
fn t36_short_not_emergency() {
    assert!(!engine().is_emergency_number("US", "999"));
}

#[test]
fn t37_short_emergency_gb() {
    assert!(engine().is_emergency_number("GB", "999"));
}

#[test]
fn t38_short_valid() {
    assert!(engine().short_is_valid("US", "911"));
}

#[test]
fn t39_short_cost() {
    assert_eq!(engine().short_expected_cost("US", "911"), pn::COST_TOLL_FREE);
}

#[test]
fn t40_short_example() {
    assert_eq!(engine().short_example_number("US"), "112");
}

#[test]
fn t41_time_zones_us() {
    assert_eq!(
        engine().time_zones_for_number("US", "2015550123"),
        vec!["America/New_York".to_string()]
    );
}

#[test]
fn t42_time_zones_gb() {
    assert_eq!(
        engine().time_zones_for_number("GB", "2070313000"),
        vec!["Europe/London".to_string()]
    );
}

#[test]
fn t43_unknown_time_zone() {
    assert_eq!(engine().unknown_time_zone(), "Etc/Unknown");
}

#[test]
fn t44_carrier_name() {
    assert_eq!(engine().carrier_name_for_number("GB", "7106000000"), "O2");
}

#[test]
fn t45_geo_description() {
    assert_eq!(
        engine().geo_description_for_number("US", "6502530000"),
        "Mountain View, CA"
    );
}

// ---- extras exercising the typed idiomatic surface ----

#[test]
fn number_type_enum_matches_raw() {
    assert_eq!(
        engine().number_type_enum("US", "2015550123"),
        pn::NumberType::FixedLineOrMobile
    );
}

#[test]
fn reason_enum_matches_raw() {
    assert_eq!(
        engine().is_possible_number_with_reason_enum("US", "201555"),
        pn::ValidationResult::TooShort
    );
}

#[test]
fn match_enum_matches_raw() {
    assert_eq!(
        engine().is_number_match_enum("+12015550123", "+1 201 555 0123"),
        pn::MatchType::Exact
    );
}

#[test]
fn source_enum_matches_raw() {
    let pn = engine();
    let num = pn.parse("+1 201 555 0123", "US");
    assert_eq!(num.source_enum(), pn::CountryCodeSource::FromNumberWithPlus);
}

#[test]
fn regions_vec_matches_count() {
    let pn = engine();
    assert_eq!(pn.regions().len() as i32, pn.region_count());
}

#[test]
fn cost_enum_matches_raw() {
    assert_eq!(
        engine().short_expected_cost_enum("US", "911"),
        pn::Cost::TollFree
    );
}

#[test]
fn free_functions_share_one_engine() {
    // The crate-level free functions load one process-wide engine.
    assert_eq!(pn::abi_version(), 6);
    assert_eq!(pn::country_code("US"), "1");
    let num = pn::parse("+1 201 555 0123", "US");
    assert_eq!(num.national_number(), "2015550123");
}
